### PR TITLE
feature:add seata-spring-boot-starter

### PR DESCRIPTION
**Pull request by xingfudeshi**
### Ⅰ. Describe what this PR did
Support spring-boot auto configuration of Seata.

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
I've done integration test with spring-boot 2.1.4.RELEASE and spring-boot 1.5.22.RELEASE

### Ⅳ. Describe how to verify it
If your project uses spring-boot, just add the following dependency to the pom.
```
<dependency>
      <groupId>io.seata</groupId>
      <artifactId>seata-spring-boot-starter</artifactId>
      <version>{seata.version}</version>
</dependency>
```
In your application.yml 
```
seata:
  application-id: test
  tx-service-group: my_test_tx_group
```
Or  application.properties
```
seata.application-id=test
seata.tx-service-group=my_test_tx_group
```
Other configurations have default values.

### CODE REVIEW

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/SpringProperties.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * @author xingfudeshi@gmail.com
+ * @date 2019/09/30
+ */
+@Component
+@ConfigurationProperties(prefix = SPRING_PREFIX)
+public class SpringProperties {
+    /**
+     * auto proxy the DataSource bean
+     */
+    private boolean datasourceAutoproxy = true;
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/SpringProperties.java]: image
the datasourceAutoproxy default is false?

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/SpringProperties.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.spring.boot.autoconfigure.properties.file;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/SpringProperties.java]: Unused import

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/SpringProperties.java (hunk 1 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/SpringProperties.java]: When using seata-spring-boot-starter,the default value will be true,Without using seata-spring-boot-starter,the dfault value is false.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/SupportProperties.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.spring.boot.autoconfigure.properties.file;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/SupportProperties.java]: Unused import

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryEurekaProperties.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import static io.seata.spring.boot.autoconfigure.StarterConstants.REGISTRY_EUREKA_PREFIX;
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/10/03
+ */
+@Component
+@ConfigurationProperties(prefix = REGISTRY_EUREKA_PREFIX)
+public class RegistryEurekaProperties {
+    private String serviceUrl = "http://localhost:1001/eureka";
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryEurekaProperties.java]: commonly used port 8761, Do you see if it is necessary to modify.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryNacosProperties.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/10/03
+ */
+@Component
+@ConfigurationProperties(prefix = REGISTRY_NACOS_PREFIX)
+public class RegistryNacosProperties {
+    private String serverAddr = "localhost";
+    private String namespace = "public";
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryNacosProperties.java]: namespace = ""?

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryEurekaProperties.java (hunk 4 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryEurekaProperties.java]: I refer to the configurations in this file https://github.com/seata/seata/blob/develop/config/seata-config-core/src/main/resources/registry.conf.
Seems some configurations are not unified in Seata.I'll modify it.Thanks!

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryNacosProperties.java (hunk 5 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryNacosProperties.java]: done.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryEurekaProperties.java (hunk 4 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryEurekaProperties.java]: done.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/StarterConstants.java [hunk 6]
```diff
@@ -130,4 +130,12 @@
     public static String NORMALIZED_KEY_CLIENT_LOCK = "client.lock.";
     public static String SPECIAL_KEY_TRANSPORT_THREAD_FACTORY = "transport.thread-factory.";
     public static String NORMALIZED_KEY_TRANSPORT_THREAD_FACTORY = "transport.thread-factory.";
+
+    public static String SPECIAL_KEY_REGISTRY_ZK="registry.zk.";
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/StarterConstants.java]: = both blank

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/provider/SpringBootConfigurationProvider.java [hunk 7]
```diff
@@ -142,22 +147,47 @@ private String convertDataId(String rawDataId) {
             String suffix = StringUtils.removeStart(rawDataId, NORMALIZED_KEY_TRANSPORT_THREAD_FACTORY);
             return StarterConstants.THREAD_FACTORY_PREFIX + "." + StringFormatUtils.minusToCamel(suffix);
         }
+        if(rawDataId.startsWith(SPECIAL_KEY_REGISTRY_ZK)){
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/provider/SpringBootConfigurationProvider.java]: `if` format like `if (rawDataId.endsWith(SPECIAL_KEY_GROUPLIST)) {`

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/StarterConstants.java (hunk 6 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/StarterConstants.java]: done.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/provider/SpringBootConfigurationProvider.java (hunk 7 above)
Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/provider/SpringBootConfigurationProvider.java]: done

File: seata-spring-boot-starter/pom.xml [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  ~  limitations under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>io.seata</groupId>
+    <artifactId>seata-spring-boot-starter</artifactId>
+    <version>0.9.0-SNAPSHOT</version>
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/pom.xml]: add parent tag??

Comment (xingfudeshi) [reply seata-spring-boot-starter/pom.xml]: > add parent tag? and the version ${revision} ?

This module and parent are independent of each other.Like seata-all module.

Comment (xingfudeshi) [reply seata-spring-boot-starter/pom.xml]: I will see if there are something i can do.

Comment (xingfudeshi) [reply seata-spring-boot-starter/pom.xml]: > add parent tag? and the version ${revision} ?

It's done.

File: seata-spring-boot-starter/pom.xml [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>seata-codec-all</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>seata-spring</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/pom.xml]: use seata-all replace???

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryRedisProperties.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/10/03
+ */
+@Component
+@ConfigurationProperties(prefix = REGISTRY_REDIS_PREFIX)
+public class RegistryRedisProperties {
+    private String serverAddr = "localhost:6379";
+    private String db = "0";
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryRedisProperties.java]: the type use int?

Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryRedisProperties.java]: The original type is string.
image

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/TransactionProperties.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+@ConfigurationProperties(prefix = TRANSACTION_PREFIX)
+public class TransactionProperties {
+    private boolean undoDataValidation = true;
+    private String undoLogSerialization = "jackson";
+    private int undoLogSaveDays = 7;
+    /**
+     * schedule delete expired undo_log in milliseconds
+     */
+    private long undoLogDeletePeriod = 86400000L;
+    private String undoLogTable = "undo_log";
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/TransactionProperties.java]: use constants? `ConfigurationKeys.TRANSACTION_UNDO_LOG_DEFAULT_TABLE`

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/TransactionProperties.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/09/30
+ */
+@Component
+@ConfigurationProperties(prefix = TRANSACTION_PREFIX)
+public class TransactionProperties {
+    private boolean undoDataValidation = true;
+    private String undoLogSerialization = "jackson";
+    private int undoLogSaveDays = 7;
```
Comment (jsbxyyx) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/TransactionProperties.java]: `UndoLogDeleteRequest.DEFAULT_SAVE_DAYS` ?

**Comment by jsbxyyx** [comment]
please check `*Properties` default value has constants. if has constants then use the constants.

**Comment by xingfudeshi** [comment]
> please check `*Properties` default value has constants. if has constants then use the constants.

sure.i will check.

File: config/seata-config-core/src/main/java/io/seata/config/ConfigurationFactory.java [hunk 13]
```diff
@@ -91,7 +99,15 @@ private static Configuration buildConfiguration() {
                 + FILE_TYPE + ConfigurationKeys.FILE_CONFIG_SPLIT_CHAR
                 + NAME_KEY;
             String name = CURRENT_FILE_INSTANCE.getConfig(pathDataId);
-            return new FileConfiguration(name);
+            Configuration configuration = new FileConfiguration(name);
+            Configuration extConfiguration=null;
+            try {
+                extConfiguration=EnhancedServiceLoader.load(ExtConfigurationProvider.class).provide(configuration);
```
Comment (zjinlei) [inline config/seata-config-core/src/main/java/io/seata/config/ConfigurationFactory.java]: code format

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/ClientProperties.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/09/30
+ */
+@Component
+@ConfigurationProperties(prefix = CLIENT_PREFIX)
+public class ClientProperties {
+    private int asyncCommitBufferLimit = 10000;
+    private int reportRetryCount = 5;
+
```
Comment (zjinlei) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/ClientProperties.java]: add(refer http://seata.io/zh-cn/docs/user/configurations.html)
1.client.lock.retry.policy.branch-rollback-on-conflict
2.client.tm.commit.retry.count
3.client.tm.rollback.retry.count

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/ServiceProperties.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    private String grouplist = "127.0.0.1:8091";
+    /**
+     * degrade current not support
+     */
+    private boolean enableDegrade = false;
+    /**
+     * disable
+     */
+    private boolean disable = false;
```
Comment (zjinlei) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/file/ServiceProperties.java]: add
service.disableGlobalTransaction

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/ConfigProperties.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/10/03
+ */
+@Component
+@ConfigurationProperties(prefix = CONFIG_PREFIX)
+public class ConfigProperties {
+    /**
+     * file、nacos 、apollo、zk
```
Comment (zjinlei) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/ConfigProperties.java]: file,nacos,apollo,zk,consul,etcd3

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryEecd3Properties.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+import static io.seata.spring.boot.autoconfigure.StarterConstants.REGISTRY_ETCD3_PREFIX;
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/10/03
+ */
+@Component
+@ConfigurationProperties(prefix = REGISTRY_ETCD3_PREFIX)
+public class RegistryEecd3Properties {
```
Comment (zjinlei) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/registry/RegistryEecd3Properties.java]: Eecd->Etcd

File: seata-spring-boot-starter/pom.xml [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        <dependencies>
+            <dependency>
+                <groupId>org.springframework.boot</groupId>
+                <artifactId>spring-boot-dependencies</artifactId>
+                <version>${spring-boot.version}</version>
+                <type>pom</type>
+                <scope>import</scope>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
```
Comment (zjinlei) [inline seata-spring-boot-starter/pom.xml]: What is this effect?

Comment (xingfudeshi) [reply seata-spring-boot-starter/pom.xml]: for using dependencies about spring-boot as we've already defined a parent.

File: pom.xml [hunk 19]
```diff
@@ -44,6 +44,7 @@
         <module>tm</module>
         <module>metrics</module>
         <module>codec</module>
+        <module>seata-spring-boot-starter</module>
```
Comment (slievrly) [inline pom.xml]: need with seata prefix？

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataAutoConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.DependsOn;
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/09/30
+ */
+@ComponentScan(basePackages = "io.seata.spring.boot.autoconfigure.properties")
+@ConditionalOnProperty(prefix = StarterConstants.SEATA_PREFIX, name = "enabled", matchIfMissing = true)
+@ConditionalOnClass({DataSourceProxy.class})
```
Comment (slievrly) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataAutoConfiguration.java]: if not contains AT mode.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataAutoConfiguration.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.DependsOn;
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/09/30
+ */
+@ComponentScan(basePackages = "io.seata.spring.boot.autoconfigure.properties")
+@ConditionalOnProperty(prefix = StarterConstants.SEATA_PREFIX, name = "enabled", matchIfMissing = true)
```
Comment (slievrly) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/SeataAutoConfiguration.java]: check seata.enable value.

**Comment by anselleeyy** [comment]
how to add seata-spring-boot-starter:1.0.0 dependency in my springboot project ? I cannot find this in mvnrepository.com

**Pull request merged**